helper/include/odp/helper/ip.h

```diff
@@ -102,49 +127,55 @@ ODP_STATIC_ASSERT(sizeof(odph_ipv4hdr_t) == ODPH_IPV4HDR_LEN,
 static inline int odph_ipv4_csum_valid(odp_packet_t pkt)
 {
 	uint32_t offset;
-	odp_u16be_t res = 0;
-	uint16_t *w;
-	int nleft = sizeof(odph_ipv4hdr_t);
+	int res;
 	odph_ipv4hdr_t ip;
-	odp_u16be_t chksum;
+	odp_u16sum_t chksum, cur_chksum;
 
 	offset = odp_packet_l3_offset(pkt);
 	if (offset == ODP_PACKET_OFFSET_INVALID)
 		return 0;
 
 	odp_packet_copy_to_mem(pkt, offset, sizeof(odph_ipv4hdr_t), &ip);
 
-	w = (uint16_t *)(void *)&ip;
 	chksum = ip.chksum;
-	ip.chksum = 0x0;
 
-	res = odph_chksum(w, nleft);
-	return (res == chksum) ? 1 : 0;
+	res = odph_ipv4_csum(pkt, offset, &ip, &cur_chksum);
+	if (odp_unlikely(res < 0))
+		return 0;
+
+	return (cur_chksum == chksum) ? 1 : 0;
 }
 
 /**
  * Calculate and fill in IPv4 checksum
  *
- * @note when using this api to populate data destined for the wire
- * odp_cpu_to_be_16() can be used to remove sparse warnings
- *
  * @param pkt  ODP packet
  *
- * @return IPv4 checksum in host cpu order, or 0 on failure
+ * @retval 0 on success
+ * @retval <0 on failure
  */
-static inline odp_u16sum_t odph_ipv4_csum_update(odp_packet_t pkt)
+static inline int odph_ipv4_csum_update(odp_packet_t pkt)
 {
-	uint16_t *w;
-	odph_ipv4hdr_t *ip;
-	int nleft = sizeof(odph_ipv4hdr_t);
+	uint32_t offset;
+	odph_ipv4hdr_t ip;
+	odp_u16sum_t chksum;
+	int res;
 
-	ip = (odph_ipv4hdr_t *)odp_packet_l3_ptr(pkt, NULL);
-	if (ip == NULL)
-		return 0;
+	offset = odp_packet_l3_offset(pkt);
+	if (offset == ODP_PACKET_OFFSET_INVALID)
+		return -1;
+
+	res = odp_packet_copy_to_mem(pkt, offset, sizeof(ip), &ip);
+	if (odp_unlikely(res < 0))
+		return res;
+
+	res = odph_ipv4_csum(pkt, offset, &ip, &chksum);
```

Comment:
Since you're deleting setting the chksum to 0 elsewhere before calling odph_ipv4_csum_update(), don't you need to set it to zero here before calling odph_ipv4_csum()? This field is supposed to be initialized to zero to calculate the checksum.

Reply:
No. `odph_ipv4_csum()` initialises checksum to 0 in the header copy. This simplifies code for header verification, etc.

Reply:
OK, I missed that on first read.